swarm/netsim/simulation.go

```diff
@@ -0,0 +1,262 @@
+	"github.com/ethereum/go-ethereum/p2p/simulations/adapters"
+)
+
+// Common errors that are returned by functions in this package.
+var (
+	ErrNodeNotFound = errors.New("node not found")
+	ErrNoPivotNode  = errors.New("no pivot node set")
+)
+
+// Package defaults.
+var (
+	DefaultHTTPSimPort = "8888"
+)
+
+// Simulation provides methods on network, nodes and services
+// to manage them.
+type Simulation struct {
+	// Net is exposed as a way to access lower level functionalities
+	// of p2p/simulations.Network.
+	Net *simulations.Network
+
+	serviceNames []string
+	cleanupFuncs []func()
+	buckets      map[discover.NodeID]*sync.Map
+	pivotNodeID  *discover.NodeID
+	shutdownWG   sync.WaitGroup
+	done         chan struct{}
+	mu           sync.RWMutex
+
+	httpSrv *http.Server        //attach a HTTP server via SimulationOptions
+	handler *simulations.Server //HTTP handler for the server
+	runC    chan struct{}       //channel where frontend signals it is ready
+}
+
+//SimulationOptions defines options for the Simulation
+//Currently only defines options to allow to attach a
+//HTTP server to the simulation (i.e. for sim frontend visualization)
+type SimulationOptions struct {
+	WithHTTP    bool   //set to true to attach a HTTP server
+	HTTPSimPort string //string defining the HTTP server port
+}
+
+// ServiceFunc is used in NewSimulation to declare new service constructor.
+// The first argument provides ServiceContext from the adapters package
+// giving for example the access to NodeID. Second argument is the sync.Map
+// where all "global" state related to the service should be kept.
+// All cleanups needed for constructed service and any other constructed
+// objects should ne provided in a single returned cleanup function.
+type ServiceFunc func(ctx *adapters.ServiceContext, bucket *sync.Map) (s node.Service, cleanup func(), err error)
+
+// NewSimulation creates a new Simulation instance with new
+// simulations.Network initialized with provided services.
+func NewSimulation(services map[string]ServiceFunc, opts *SimulationOptions) (s *Simulation) {
+	s = &Simulation{
+		buckets: make(map[discover.NodeID]*sync.Map),
+		done:    make(chan struct{}),
+	}
+
+	adapterServices := make(map[string]adapters.ServiceFunc, len(services))
+	for name, serviceFunc := range services {
+		s.serviceNames = append(s.serviceNames, name)
+		adapterServices[name] = func(ctx *adapters.ServiceContext) (node.Service, error) {
+			b := new(sync.Map)
+			service, cleanup, err := serviceFunc(ctx, b)
+			if err != nil {
+				return nil, err
+			}
+			s.mu.Lock()
+			defer s.mu.Unlock()
+			if cleanup != nil {
+				s.cleanupFuncs = append(s.cleanupFuncs, cleanup)
+			}
+			s.buckets[ctx.Config.ID] = b
+			return service, nil
+		}
+	}
+
+	s.Net = simulations.NewNetwork(
+		adapters.NewSimAdapter(adapterServices),
+		&simulations.NetworkConfig{ID: "0"},
+	)
+
+	if opts == nil {
+		opts = new(SimulationOptions)
+	}
+	if opts.WithHTTP {
```

Comment:
consider putting http server related stuff in a separate file `server.go` or `http.go`

Reply:
@holisticode http server related things are now in separate file and test also, but I did not address other @zelig comments regarding http server

Reply:
By other comments you mean the "WithServer" pattern?

Reply:
This one too https://github.com/ethersphere/go-ethereum/pull/769#discussion_r199889140